### PR TITLE
fix(api): fix refresh token withings

### PR DESCRIPTION
**Pull request by Orta21**
Ticket: #[313](https://github.com/metriport/metriport-internal/issues/313)

### Description

Refresh token error withings

### Release Plan

ASAP

### CODE REVIEW

**Comment by Orta21** [comment]
We discussed expires_at across the providers. Thats not included in the PR since its mostly just the withings stuff plus the access token logic change but i can add it in here if we dont want to create another ticket and PR

File: api/app/src/providers/withings.ts [hunk 1]
```diff
@@ -79,6 +76,60 @@ export class Withings extends Provider implements OAuth2 {
...
+    const providerData = getProviderDataFromConnectUserOrFail(connectedUser, PROVIDER_WITHINGS);
+
+    const token = providerData.token;
+
+    const refreshedToken = await this.checkRefreshToken(token, connectedUser);
+
+    return refreshedToken.access_token;
+  }
+
+  async checkRefreshToken(token: string, connectedUser: ConnectedUser): Promise<Token> {
```
Comment (Orta21) [inline api/app/src/providers/withings.ts:89]: The getAccessToken specifically for withings could be cleaner and refactored with whats in oauth2 just trying to push it out quickly but let me know and i can clean it up

File: api/app/src/providers/withings.ts [hunk 2]
```diff
@@ -79,6 +76,60 @@ export class Withings extends Provider implements OAuth2 {
...
+    const access_token = JSON.parse(token)
+    const now = dayjs(Date.now()).unix()
+    const isExpired = access_token.expires_at < now;
```
Comment (Goncharo) [inline api/app/src/providers/withings.ts:92]: Should we use the same logic as the simple oauth2 lib? They use an expiration window so the expiry is caught a bit earlier

**Review by Goncharo — CHANGES_REQUESTED** [review]
Small comment, also should we move the retrieval of the token for all providers? We only have whoop/Fitbit here.

File: api/app/src/providers/withings.ts (hunk 1 above)
Comment (Goncharo) [reply api/app/src/providers/withings.ts:89]: All good for now

File: api/app/src/providers/withings.ts [hunk 3]
```diff
@@ -79,6 +76,60 @@ export class Withings extends Provider implements OAuth2 {
...
+
+    const token = providerData.token;
+
+    const refreshedToken = await this.checkRefreshToken(token, connectedUser);
+
+    return refreshedToken.access_token;
+  }
+
+  async checkRefreshToken(token: string, connectedUser: ConnectedUser): Promise<Token> {
+    const access_token = JSON.parse(token)
```
Comment (Goncharo) [inline api/app/src/providers/withings.ts:90]: missing `;` - why isn't the linting catching this?

File: api/app/src/providers/withings.ts [hunk 4]
```diff
@@ -79,6 +76,60 @@ export class Withings extends Provider implements OAuth2 {
...
+          "https://wbsapi.withings.net/v2/oauth2",
+          `action=requesttoken&grant_type=refresh_token&client_id=${Withings.clientId}&client_secret=${Withings.clientSecret}&refresh_token=${access_token.refresh_token}`,
+          {
+            headers: {
+              "Content-Type": "application/x-www-form-urlencoded",
+            },
+          }
+        );
+
+        response.data.body.expires_at = dayjs(Date.now()).add(response.data.body.expires_in, "seconds").unix()
```
Comment (Goncharo) [inline api/app/src/providers/withings.ts:105]: missing ;

File: api/app/src/providers/withings.ts [hunk 5]
```diff
@@ -79,6 +76,60 @@ export class Withings extends Provider implements OAuth2 {
...
+
+        await updateProviderData({
+          id: connectedUser.id,
+          cxId: connectedUser.cxId,
+          provider: PROVIDER_WITHINGS,
+          providerItem,
+        });
+
+        return response.data.body;
+        1674682301644
```
Comment (Goncharo) [inline api/app/src/providers/withings.ts:122]: leftover?

File: api/app/src/providers/withings.ts [hunk 6]
```diff
@@ -79,6 +76,60 @@ export class Withings extends Provider implements OAuth2 {
...
+
+      } catch (error) {
+        console.log("Error refreshing access token: ", error);
+        throw new Error("Error refreshing access token: ");
+      }
+    }
+
+    return access_token
```
Comment (Goncharo) [inline api/app/src/providers/withings.ts:130]: missing ;

**Review by Goncharo — CHANGES_REQUESTED** [review]

**Review by Goncharo — APPROVED** [review]
🚢

**Pull request merged**